### ISSUE TITLE
'declare method' quick fix for adding a private method

### ISSUE THREAD

**Issue by mjbvz** [microsoft/TypeScript#37782]
From https://github.com/microsoft/vscode/issues/94118

**TypeScript Version:**  3.9.0-dev.20200330

**Search terms**
- quick fix
- declare method
- private

**Repro**
For ts file:

`index.ts`
```ts
class Bar {
    bar() {
         this._baz(123)
    }
}
```

1. Trigger quick fixes on `_baz`

**Feature request:**
https://github.com/microsoft/TypeScript/issues/36249 added a quick fix for adding private properties. However there is no quick fix for adding a private method

**Playground Link:** 

**Related Issues:** 
- https://github.com/microsoft/TypeScript/issues/36249

**Comment by bpasero** [comment]
I am the original one complaining about this: what I expect is NOT having 1 code action per visibility, but rather ONE code action (as today) with a tab-stop at the visibility that I can flip the various visibilities. I realise this may need support from VSCode too. Here is how it would work:
* you trigger the code action
* you get into a mode like snippets where certain places are stops you can cycle through with `Tab` key
* each method or property has a tab stop where you can navigate to and once you are there, be able to select a modifier `private`, `protected`, `public` etc.

**Comment by a-tarasyuk** [comment]
@DanielRosenwasser Is there a way to have a sub action to handle _`tab-stop`_? Or need to generate all possible actions for various methods/properties and VSCode should resolve _`tab-stop`_? Do I need to close https://github.com/microsoft/TypeScript/pull/37806, and revert https://github.com/microsoft/TypeScript/issues/36249?

**Comment by DanielRosenwasser** [comment]
Could you explain what you mean by *`tab-stop`*? I feel like I have an idea, but I might be out of the loop on something.

**Comment by a-tarasyuk** [comment]
@bpasero 
> Could you explain what you mean by tab-stop?

**Comment by mjbvz** [comment]
Tabstips are from VS Code's snippets: https://code.visualstudio.com/docs/editor/userdefinedsnippets#_tabstops

There is no concept of tabstops in the TS Server api today since it only talks about text. I've opened #25207 which proposes adding the idea of snippets to typescript

**Comment by a-tarasyuk** [comment]
Right now I see two ways for this issue
1. Resolve this issue based on the original requirements 
2. Close #37806 revert #36249 and add _`tab-stop`_ for modifiers after #25207 will be implemented

> I feel like I have an idea

@DanielRosenwasser Maybe you have a better idea :)

**Comment by DanielRosenwasser** [comment]
Hah, I meant like "I might have an idea of what you mean by tab stops". I think the weird thing is that the rename location already allows you to make something `#`-private, but it's not clear whether we could make a conditional tabstop location for when the declaration *isn't* `#`-private

**Comment by bpasero** [comment]
To demonstrate what I expect to happen, here is what Eclipse does:

![kap](https://user-images.githubusercontent.com/900690/81466136-7aa34680-91cf-11ea-83d6-2f03bcbac131.gif)

And what VSCode does:

![kap](https://user-images.githubusercontent.com/900690/81466151-91499d80-91cf-11ea-9cb4-e1a4c3e8d25b.gif)

Paper cuts:
* method is added above constructor???
* no way to change anything after the method was inserted

Can we reopen this or reconsider this?

**Comment by sandersn** [comment]
Yeah, #37806 partially fixes this but doesn't address the full issue. I'll re-open that so we can unify the two fixes when post-codefix rename locations are available.

**Comment by asd34-45** [comment]
> From [microsoft/vscode#94118](https://github.com/microsoft/vscode/issues/94118)
> 
> **TypeScript Version:** 3.9.0-dev.20200330
> 
> **Search terms**
> 
> * quick fix
> * declare method
> * private
> 
> **Repro**
> For ts file:
> 
> `index.ts`
> 
> ```ts
> class Bar {
>     bar() {
>          this._baz(123)
>     }
> }
> ```
> 
> 1. Trigger quick fixes on `_baz`
> 
> **Feature request:**
> #36249 added a quick fix for adding private properties. However there is no quick fix for adding a private method
> 
> **Playground Link:**
> 
> **Related Issues:**
> 
> * #36249
echo off@$

**Comment by asd34-45** [comment]
![Screenshot_2020-05-29-01-13-45-89_68e74cc0265513bba3f4cf8e2c9891fe](https://user-images.githubusercontent.com/66072261/83342679-d6f31500-a2fa-11ea-8968-135270a9e3d3.png)

**Comment by asd34-45** [comment]
192.168.1.1

**Comment by arvind-nikam-halodoc** [comment]
> From [microsoft/vscode#94118](https://github.com/microsoft/vscode/issues/94118)
> 
> **TypeScript Version:** 3.9.0-dev.20200330
> 
> **Search terms**
> 
> * quick fix
> * declare method
> * private
> 
> **Repro**
> For ts file:
> 
> `index.ts`
> 
> ```ts
> class Bar {
>     bar() {
>          this._baz(123)
>     }
> }
> ```
> 
> 1. Trigger quick fixes on `_baz`
> 
> **Feature request:**
> #36249 added a quick fix for adding private properties. However there is no quick fix for adding a private method
> 
> **Playground Link:**
> 
> **Related Issues:**
> 
> * #36249

It’s fixed i guess.
<img width="596" alt="Screenshot 2021-03-10 at 11 26 37 AM" src="https://user-images.githubusercontent.com/79127332/110583455-bb733f80-8193-11eb-82f6-1864e45ed89f.png">

**Comment by NsdHSO** [comment]
@DanielRosenwasser is it still open?

**Comment by kamolovd** [comment]
This is my first investment in outsourcing, can I start with this? @mjbvz

**Comment by tysoncung** [comment]
I noticed this issue. Have you tried checking the error logs or console output? That might help narrow down the root cause. I'd be happy to help investigate if you can share more details about your environment (OS, version, etc.).

**Comment by oppong07** [comment]
Hi! I’m going to implement the “declare method” quick fix and will open a PR from branch feat/37782-declare-method-quick-fix (fork: oppong07/TypeScript). I’ll add tests demonstrating the failing diagnostic and the expected result after applying the fix, and I will link this issue from the PR.
If maintainers prefer this work be submitted to microsoft/typescript-go instead, please let me know.
Branch: https://github.com/oppong07/TypeScript/tree/feat/37782-declare-method-quick-fix

**Comment by tejash5489-lang** [comment]
This looks like it's already resolved for the originally-reported case. #37806 (merged 2020-05-06) added the "Declare private method" / "Declare method" quick-fix actions in `fixAddMissingMember.ts`, and there's a fourslash test (`codeFixAddMissingMember18_declarePrivateMethod.ts`) covering exactly the repro from this issue:

```ts
class A {
    constructor() {
        this._foo();
    }
}
```

Triggering the quick fix on `_foo` today produces `private _foo() { throw new Error("Method not implemented."); }`, and `_baz(123)` from the original repro would work the same way — TS infers `private` from the leading underscore via `startsWithUnderscore()`.

What's still open is the follow-up scope from later in this thread: a single linked quick-fix with a tab-stop to cycle between `private`/`protected`/`public` (and rename the call site in sync), which needs snippet/tabstop support in the language service — tracked separately in #25207 and still unimplemented. That's a materially bigger change than "add a declare-method codefix," so it's probably worth relabeling or splitting into its own issue so this one can close and new contributors don't duplicate the already-merged #37806 work (as looks like it might be about to happen).

**Comment by LeonxLJX** [comment]
I'd like to take this one — I'll follow up with a PR. (claiming via @LeonxLJX)

**Comment by LeonxLJX** [comment]
Hi! I'd like to add the 'declare method' quick fix. Plan: implement the code fix, add tests. May I be assigned?
